internal/router/controllers/system.go

```diff
@@ -83,21 +81,117 @@ func (c *SystemController) GetTablesByController(rw http.ResponseWriter, r *http
 	// Not doing `omitempty` in tableland.Table since
 	// that feels hacky. Looks safer to define a separate type here at the handler level.
 	type tableNameIDUnified struct {
-		Controller string    `json:"controller"`
-		Name       string    `json:"name"`
-		Structure  string    `json:"structure"`
-		CreatedAt  time.Time `json:"created_at"`
+		Controller string `json:"controller"`
+		Name       string `json:"name"`
+		Structure  string `json:"structure"`
 	}
 	retTables := make([]tableNameIDUnified, len(tables))
 	for i, t := range tables {
 		retTables[i] = tableNameIDUnified{
 			Controller: t.Controller,
-			Name:       fmt.Sprintf("%s_%d_%s", t.Prefix, t.ChainID, t.ID),
+			Name:       t.Name(),
 			Structure:  t.Structure,
-			CreatedAt:  t.CreatedAt,
 		}
 	}
 
 	rw.WriteHeader(http.StatusOK)
 	_ = json.NewEncoder(rw).Encode(retTables)
 }
+
+// GetTablesByStructureHash handles the GET /chain/{id}/tables/structure/{hash} call.
+func (c *SystemController) GetTablesByStructureHash(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	rw.Header().Set("Content-type", "application/json")
+	vars := mux.Vars(r)
+
+	hash := vars["hash"]
+	tables, err := c.systemService.GetTablesByStructure(ctx, hash)
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		log.Ctx(ctx).
+			Error().
+			Err(err).
+			Str("hash", hash).
+			Msg("failed to fetch tables")
+
+		_ = json.NewEncoder(rw).Encode(errors.ServiceError{Message: "Failed to fetch tables"})
+		return
+	}
+
+	type tableNameIDUnified struct {
+		Controller string `json:"controller"`
+		Name       string `json:"name"`
+		Structure  string `json:"structure"`
+	}
+	retTables := make([]tableNameIDUnified, len(tables))
+	for i, t := range tables {
+		retTables[i] = tableNameIDUnified{
+			Controller: t.Controller,
+			Name:       t.Name(),
+			Structure:  t.Structure,
+		}
+	}
+
+	rw.WriteHeader(http.StatusOK)
+	_ = json.NewEncoder(rw).Encode(retTables)
+}
+
+// GetSchemaByTableName handles the GET /schema/{table_name} call.
+func (c *SystemController) GetSchemaByTableName(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	rw.Header().Set("Content-type", "application/json")
+	vars := mux.Vars(r)
+
+	name := vars["table_name"]
+	schema, err := c.systemService.GetSchemaByTableName(ctx, name)
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		log.Ctx(ctx).
+			Error().
+			Err(err).
+			Str("name", name).
+			Msg("failed to fetch tables")
+
+		_ = json.NewEncoder(rw).Encode(errors.ServiceError{Message: "Failed to get schema from table"})
+		return
+	}
+
+	if len(schema.Columns) == 0 {
```

Comment:
This is a check for empty schema. Unfortunately, `if (sqlstore.Schema{}) == schema` is not an option because the fields of `sqlstore.Schema` are not comparable.